torch/_functorch/aot_autograd.py

```diff
@@ -1860,22 +1860,43 @@ def backward(ctx, *flat_args):
                 list(ctx.symints) + list(ctx.saved_tensors) + list(contiguous_args)
             )
             del contiguous_args
-            if CompiledFunction.compiled_bw is None:
-                # TODO - pass in fake tensors ?
-                context = disable_autocast_manager if disable_amp else nullcontext
-                with context(), track_graph_compiling(aot_config, "backward"):
-                    CompiledFunction.compiled_bw = aot_config.bw_compiler(
-                        bw_module, all_args
-                    )
 
-            ctx.maybe_clear_saved_tensors()
-            out = call_func_with_args(
-                CompiledFunction.compiled_bw,
-                all_args,
-                steal_args=True,
-                disable_amp=disable_amp,
-            )
-            return tuple(out)
+            def call_compiled_backward():
+                if CompiledFunction.compiled_bw is None:
+                    # TODO - pass in fake tensors ?
+                    context = disable_autocast_manager if disable_amp else nullcontext
+                    with context(), track_graph_compiling(aot_config, "backward"):
+                        CompiledFunction.compiled_bw = aot_config.bw_compiler(
+                            bw_module, all_args
+                        )
+
+                ctx.maybe_clear_saved_tensors()
+                out = call_func_with_args(
+                    CompiledFunction.compiled_bw,
+                    all_args,
+                    steal_args=True,
+                    disable_amp=disable_amp,
+                )
+
+                return tuple(out)
+
+            if torch.is_grad_enabled() and any(t.requires_grad for t in all_args if isinstance(t, torch.Tensor)):
+                # Ensure that the graph is connected, and error if double backward is performed.
+                # See comment for why once_differentiable is not sufficient:
+                # https://github.com/pytorch/pytorch/pull/92348/files#r1072962107
+                class CompiledFunctionBackward(torch.autograd.Function):
+                    @staticmethod
+                    def forward(ctx, *unused_args):
+                        return call_compiled_backward()
```

Comment:
does autograd mess with the arguments at all between `.apply(*all_args)` and `.forward(*unused_args)` that require us to use the passed-in args instead of the originals 🤔. Off the top of my head, i'm wondering if it e.g. ever takes a view of one of the inputs, and passes that into the forward instead or something.

Just thinking out loud, but if this passes CI then it seems ok to me.

Reply:
The processing it does it here is basically the same as what VariableType does, e.g. check which arguments require grad, if any require grad, etc. It also creates a new tuple for some reason, so you'd have an extra reference from that, in any case it shouldn't matter for us if we're going down this path since the user would've wanted to keep saved variables alive for double backward anyway. This might be unrelated, but there are some cases on the other hand where autograd Function would take a view of the outputs, e.g. if you returned an output as-is.